### PR TITLE
Implement password policy caching in auth object (#2)

### PR DESCRIPTION
**Pull request by ch5zzy**
Implement password policy cache storage in AuthInternal, with methods for getting and updating the policy

### CODE REVIEW

File: packages/auth/src/core/auth/auth_impl.ts [hunk 1]
```diff
@@ -422,6 +427,41 @@ export class AuthImpl implements AuthInternal, _FirebaseService {
...
+    if (this.tenantId == null) {
+      return this._projectPasswordPolicy;
+    } else {
+      return this._tenantPasswordPolicies[this.tenantId];
+    }
+  }
+
+  async _updatePasswordPolicy(): Promise<void> {
+    const response = await _getPasswordPolicy(this);
+
```
Comment (Xiaoshouzi-gh) [inline packages/auth/src/core/auth/auth_impl.ts:440]: can you add a comment here?

File: packages/auth/src/core/auth/auth_impl.ts [hunk 2]
```diff
@@ -422,6 +427,41 @@ export class AuthImpl implements AuthInternal, _FirebaseService {
...
+    ) {
+      return Promise.reject(
+        this._errorFactory.create(
+          AuthErrorCode.PASSWORD_POLICY_VERSION_MISMATCH,
+          {}
+        )
+      );
+    }
+
+    const passwordPolicy = {
```
Comment (Xiaoshouzi-gh) [inline packages/auth/src/core/auth/auth_impl.ts:452]: const passwordPolicy:PasswordPolicy = { ...}

**Review by Xiaoshouzi-gh — CHANGES_REQUESTED** [review]

File: common/api-review/auth.api.md [hunk 3]
```diff
@@ -562,6 +562,31 @@ export interface ParsedToken {
...
+    };
+}
+
+// @public
+export interface PasswordValidationStatus {
+    readonly containsLowercaseLetter?: boolean;
+    readonly containsNonAlphanumericCharacter?: boolean;
+    readonly containsNumericCharacter?: boolean;
+    readonly containsUppercaseLetter?: boolean;
+    readonly isValid: boolean;
```
[the remark this answers is not in the document]
Comment (ch5zzy) [reply common/api-review/auth.api.md:584]: I ran `yarn docgen` to generate this, so I'm not sure why isValid is not the first property here either. Maybe someone else can comment on this.

File: packages/auth/src/core/auth/auth_impl.ts (hunk 1 above)
Comment (ch5zzy) [reply packages/auth/src/core/auth/auth_impl.ts:440]: Done.

File: packages/auth/src/core/auth/auth_impl.ts [hunk 4]
```diff
@@ -422,6 +427,41 @@ export class AuthImpl implements AuthInternal, _FirebaseService {
     }
   }
 
+  _getPasswordPolicy(): PasswordPolicy | null {
+    if (this.tenantId == null) {
```
[the remark this answers is not in the document]
Comment (ch5zzy) [reply packages/auth/src/core/auth/auth_impl.ts]: Done.

File: packages/auth/src/core/auth/auth_impl.ts (hunk 2 above)
Comment (ch5zzy) [reply packages/auth/src/core/auth/auth_impl.ts]: Done.

**Review by Xiaoshouzi-gh — APPROVED** [review]
LGTM. Please tag me in the conversion with the auto-generated file property order issue when there are updates.

File: common/api-review/auth.api.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (prameshj) [reply common/api-review/auth.api.md:584]: I suspect it orders them alphabetically..

File: packages/auth/src/api/password_policy/get_password_policy.test.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    );
+  });
+
+  it('should handle errors', async () => {
+    mockEndpoint(
+      Endpoint.GET_PASSWORD_POLICY,
+      {
+        error: {
+          code: 400,
+          message: ServerError.INVALID_PROVIDER_ID,
```
Comment (prameshj) [inline packages/auth/src/api/password_policy/get_password_policy.test.ts:74]: super nit: let's use a different error code like TOO_MANY_ATTEMPTS_TRY_LATER, which is more possible to hit in a real scenario?

File: packages/auth/src/core/auth/auth_impl.ts [hunk 6]
```diff
@@ -422,6 +427,42 @@ export class AuthImpl implements AuthInternal, _FirebaseService {
...
+      return this._projectPasswordPolicy;
+    } else {
+      return this._tenantPasswordPolicies[this.tenantId];
+    }
+  }
+
+  async _updatePasswordPolicy(): Promise<void> {
+    const response = await _getPasswordPolicy(this);
+
+    // Check that the policy schema version is supported by the SDK.
```
Comment (prameshj) [inline packages/auth/src/core/auth/auth_impl.ts:441]: nit: also add a comment like // TODO - Update this logic to use a MAX_SUPPORTED_PASSWORD_POLICY_SCHEMA_VERSION once we have multiple schema versions.

**Review by prameshj — APPROVED** [review]

File: packages/auth/src/core/auth/auth_impl.ts (hunk 6 above)
Comment (ch5zzy) [reply packages/auth/src/core/auth/auth_impl.ts:441]: Done.

File: packages/auth/src/api/password_policy/get_password_policy.test.ts (hunk 5 above)
Comment (ch5zzy) [reply packages/auth/src/api/password_policy/get_password_policy.test.ts]: Done.

File: common/api-review/auth.api.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ch5zzy) [reply common/api-review/auth.api.md:584]: Ah okay, it's interesting that the properties in `customStrengthOptions` are not ordered alphabetically. Maybe this is not enforced for nested properties.

**Pull request merged**